src/Adaptivity/RefinementRules.jl

```diff
@@ -0,0 +1,163 @@
+struct GenericRefinement <: RefinementRuleType end
+struct WithoutRefinement <: RefinementRuleType end
+
+struct RefinementRule{T,P,A,B,C}
+  poly         :: P
+  ref_grid     :: A
+  f2c_cell_map :: B
+  x_to_cell    :: C
+
+  function RefinementRule{T}(poly         ::Polytope,
+                             ref_grid     ::Grid,
+                             f2c_cell_map ::AbstractVector,
+                             x_to_cell    ::Function) where T <: RefinementRuleType
+    P = typeof(poly)
+    A = typeof(ref_grid)
+    B = typeof(f2c_cell_map)
+    C = typeof(x_to_cell)
+    new{T,P,A,B,C}(poly,ref_grid,f2c_cell_map,x_to_cell)
+  end
+end
+
+ReferenceFEs.get_polytope(rr::RefinementRule) = rr.poly
+get_ref_grid(rr::RefinementRule) = rr.ref_grid
+num_subcells(rr::RefinementRule) = num_cells(rr.ref_grid)
+num_ref_faces(rr::RefinementRule,d::Int) = num_faces(rr.ref_grid,d)
+RefinementRuleType(::RefinementRule{T}) where T = T
+
+function Geometry.get_cell_map(rr::RefinementRule)
+  return rr.f2c_cell_map
+end
+
+function get_inverse_cell_map(rr::RefinementRule)
+  return lazy_map(Fields.inverse_map,rr.f2c_cell_map)
+end
+
+function get_cell_measures(rr::RefinementRule)
+  ref_grid  = get_ref_grid(rr)
+  ref_trian = Triangulation(UnstructuredDiscreteModel(ref_grid))
+  measures  = get_cell_measure(ref_trian)
+  M = sum(measures)
+  measures /= M
+  return measures
+end
+
+function RefinementRule(::T,poly::Polytope, ref_grid::Grid; cell_maps=nothing, x_to_cell_id=nothing) where T <: RefinementRuleType
+  if (cell_maps === nothing)
+    f2c_cell_map = Geometry.get_cell_map(ref_grid)
+  else
+    @check length(cell_maps) == num_cells(ref_grid)
+    f2c_cell_map = cell_maps
+  end
+
+  if x_to_cell_id === nothing
+    ref_trian    = Triangulation(UnstructuredDiscreteModel(ref_grid))
+    p2c_cache    = CellData._point_to_cell_cache(CellData.KDTreeSearch(),ref_trian)
+    x_to_cell(x) = CellData._point_to_cell!(p2c_cache, x)
+  else
+    x_to_cell = x_to_cell_id
+  end
+
+  return RefinementRule{T}(poly,ref_grid,f2c_cell_map,x_to_cell)
+end
+
+# GenericRefinement Rule
+
+function RefinementRule(reffe::LagrangianRefFE{D},nrefs::Integer;kwargs...) where D
+  partition = tfill(nrefs,Val{D}())
+  return RefinementRule(get_polytope(reffe),partition;kwargs...)
+end
+
+function RefinementRule(reffe::LagrangianRefFE{D},partition::NTuple{D,Integer};kwargs...) where D
+  return RefinementRule(get_polytope(reffe),partition;kwargs...)
+end
+
+function RefinementRule(poly::Polytope{D},nrefs::Integer;kwargs...) where D
+  partition = tfill(nrefs,Val{D}())
+  return RefinementRule(poly,partition;kwargs...)
+end
+
+function RefinementRule(poly::Polytope{D},partition::NTuple{D,Integer};kwargs...) where D
+  ref_grid = UnstructuredGrid(compute_reference_grid(poly,partition))
+  return RefinementRule(GenericRefinement(),poly,ref_grid;kwargs...)
+end
+
+# FineToCoarseFields
+
+struct FineToCoarseField{A<:AbstractArray{<:Field},B<:RefinementRule} <: Field
```

Comment:
I think that with generlization in mind, we should also support the case in which `A<:AbstractArray{AbstractArray{<:Field}}`. 

This will be needed, e.g., when we want to transform a `SingleFieldFEBasis` defined in the child mesh to its parent. 

Does this make sense?

Reply:
why this data type is in the same source file as `RefinementRules.jl`. Shouldn't it go to its own source file? If you took at the source code of Gridap, you will see that we have separated files for each subtype of `Map` in many cases.

Reply:
@JordiManyer I will try to implement https://github.com/gridap/Gridap.jl/pull/838/files#r1042880589

Reply:
>  @JordiManyer I will try to implement https://github.com/gridap/Gridap.jl/pull/838/files#r1042880589

For the records. I have tried to implement this feature. After giving to it a quite deep though, I have arrived to the conclusion that, at least given how I conceive this operation, we do not have sufficient expressivity to implement it. 
I have came up with a hack (see below), which is clearly not acceptable. Let us thus resolve this conversation, I am not sure if we actually need this feature for GMG preconditioners/AMR. I needed it for another context, neural networks test functions, etc., the good news is that I have came up with another idea which does not actually involve all this, but the definition of a new type of FESpace that linearizes a high order space. Anyway, this dev is not for this pr. I will do it in a separate branch, and then merge if we believe that it can be useful for gridap users. 

Let me give a little bit more of context. Assume that we have a single coarse quad (coarse mesh), and its refinement into 4 children cells (fine mesh). Then, we have a bilinear h1-conforming FE space atop the fine mesh. If we now consider a FEBasis of this space, we have 4 shape functions per (children) cell, but when you transfer them to the coarse-grid cell you want to glue them up into 9 shape functions, so that the evaluation of the transformed basis into a set of coarse-grid quad points returns a 2D array, and not something a 1D array of 1D arrays (with the first index being the quadrature point id, and the second the shape function id of the child to where the quadrature point lays). 

In order to perform this gluing operation, we need access to the FESpace. And we DO NOT have it at this level. Thus, I came with the hack below (assumptions are described as comments). 


```julia
function Geometry.return_cache(a::FineToCoarseField{<:AbstractArray{<:AbstractArray{<:Field}}},x::AbstractArray{<:Point})
  fields, x_to_cell = a.fine_fields, a.rrule.x_to_cell
  cmaps = get_inverse_cell_map(a.rrule)

  xi_cache = array_cache(x)
  fi_cache = array_cache(fields)
  mi_cache = array_cache(cmaps)

  xi = getindex!(xi_cache,x,1)
  child_id = x_to_cell(xi)
  mi = getindex!(mi_cache,cmaps,child_id)
  fi = getindex!(fi_cache,fields,child_id)

  zi_cache = Fields.return_cache(mi,xi)
  zi = evaluate!(zi_cache,mi,xi)

  yi_type  = Fields.return_type(fi,zi)
  yi_cache = Fields.return_cache(fi,zi)
  y_cache  = Arrays.CachedArray(eltype(yi_type),2)

  return fi_cache, mi_cache, xi_cache, zi_cache, yi_cache, y_cache
end

function Geometry.evaluate!(cache,a::FineToCoarseField{<:AbstractArray{<:AbstractArray{<:Field}}},x::AbstractArray{<:Point})
  fi_cache, mi_cache, xi_cache, zi_cache, yi_cache, y_cache = cache
  fields, x_to_cell = a.fine_fields, a.rrule.x_to_cell
  cmaps = get_inverse_cell_map(a.rrule)

  # DIRTY Hacks:
  #  (1) Assumes TestBasis() BasisStyle of a.fields
  #  (2) Assumes a one-to-one matching among DoFs and rrule geo nodes!

  #  (1) and (2) are caused because we miss the information required to 
  #  do it cleanly/generally. For (2) we require the l2g mapping among DoF ids
  #  from the FESpace associated to the FEBasis. 
  Arrays.setsize!(y_cache, (length(x),Geometry.num_nodes(a.rrule.ref_grid)))
  
  y_cache.array .= 0.0 

  cell_node_ids = Geometry.get_cell_node_ids(a.rrule.ref_grid)
  for i in eachindex(x)
     xi = getindex!(xi_cache,x,i)
     child_id = x_to_cell(xi)
     fi = getindex!(fi_cache,fields,child_id)
     mi = getindex!(mi_cache,cmaps,child_id)
     zi = Fields.evaluate!(zi_cache,mi,xi)
     y_cache.array[i,cell_node_ids[child_id]] = Fields.evaluate!(yi_cache,fi,zi)
  end
  return y_cache.array
end
```

Code that stresses this hack:

```julia
using Gridap
using FillArrays

modelH=CartesianDiscreteModel((0,1,0,1),(1,1))
modelh=Gridap.Adaptivity.refine(modelH,2)

order  = 2
reffeH = ReferenceFE(lagrangian,Float64,order)
VH     = TestFESpace(modelH,reffeH; conformity=:H1)
UH     = TrialFESpace(VH)

order  = 1
reffeh = ReferenceFE(lagrangian,Float64,order)
Vh     = TestFESpace(modelh,reffeh; conformity=:H1)
Uh     = TrialFESpace(Vh)

duH=get_trial_fe_basis(UH)
dvh=get_fe_basis(Vh)

Ωh   = Triangulation(modelh)
dΩh  = Measure(Ωh,2*order)
ΩH   = Triangulation(modelH)
dΩH  = Measure(ΩH,2*order)

xH=Gridap.CellData.get_cell_points(dΩH.quad)

a=Gridap.CellData.change_domain(duH,Ωh,Gridap.CellData.ReferenceDomain())
ad=Gridap.CellData.get_data(a)

x=Gridap.CellData.CellPoint(Fill([Point(0.0,0.0)],4),Ωh,Gridap.CellData.ReferenceDomain())
a_dvh=a*dvh
a_dvh_x=a_dvh(x)

b=Gridap.CellData.change_domain(dvh,ΩH,Gridap.CellData.ReferenceDomain())
#xH=Gridap.CellData.CellPoint(Fill([Point(0.0,0.5)],1),ΩH,Gridap.CellData.ReferenceDomain())
b_xH=b(xH)
b_xH[1]

(duH*b)(xH)[1]





duH_xH=duH(xH)

b*duH

lazy_map(evaluate, ad, Fill([Point(0.0,0.0)],4) )[1]
lazy_map(evaluate, ad, Fill([Point(0.0,0.0)],4) )[2]
lazy_map(evaluate, ad, Fill([Point(0.0,0.0)],4) )[3]
lazy_map(evaluate, ad, Fill([Point(0.0,0.0)],4) )[4]

dΩHh=Measure(ΩH,Ωh,CellQuadrature(Ωh,2*order+1))
a(UH,vh)=∫(UH*vh)dΩHh
assemble_matrix(a,UH,Vh)
```

Reply:
I see... The only way I see this information being available is through the `RefinementRule`. However I think we want to keep this structure 100% geometrical, which does not allow it. We can write this as an issue and maybe address it a a later stage.

Reply:
In any case this probably belongs to a new PR.